### DISCUSSION TITLE
How to transfer large SB message by software bus service?

### DISCUSSION THREAD

**Discussion by adamlin09** [nasa/cFS#325] [Q&A] [answered]
Hello everyone, 

In my application, I would like to use software bus to send a large message(such as pixel data of an image), which is larger than the size that can be specified in CCSDS in theory(65536 bytes, if I understand correctly). The API, CFE_SB_InitMsg(), also seems to only create a message which is not larger than 65536 bytes. So, I can't send the message in one packet.

I want to know if there are any ways to send a large message in one packet? If not, how can I send the large message in sequence by using sequence flag and sequence count in CCSDS?

Thank you for your attention.

**Comment by ThibFrgsGmz** [comment]
I think part of the answer can be found in the FAQ in the April 2004 Draft Green Book version of the CCSDS Space Packet Protocol on page 25 (https://cwe.ccsds.org/sis/docs/Work Completed (Closed WGs)/Space Packet Protocol Working Group/Draft Documents/Space Packet Protocol (Draft Green Book, Issue 2).pdf).

I did not check again if what it says is still valid, but let me summarize the answer here:

---

In the case of sending large data buffer (image or memory data) that does not fit into a single space packet, these application data units must be transferred with a group of space packets.

The source user application must segment the application data unit into smaller segments and ensure that each segment fits into a space packet.

The Space Packet has fields called Sequence Flags in its header to identify the first and last segment in a group, and reconstruction of the original application data unit at the destination is possible using these flags.

If the segment number of each segment is to be transferred with the segment itself, the secondary header of the packet can be used to send the segment number.

---

**Reply by adamlin09** [reply]
Hi @ThibFrgsGmz ,

Thank you for replying! 

I know that CCSDS describes a sequence feature for multiple packets. But, there aren't any functions in cFS I can see for interacting with the sequence count and sequence flags in the message header. So I don't know if cFS is actually built for sending multi-packet messages in any way, or if that's a feature that CCSDS describes but cFS hasn't implemented yet.

If cFS hasn't implemented the for sending multiple messages yet, Does it mean that I should modify the message header manually and still use CFE_SB_SendMsg() function many times for sending multiple sequence packets?

**Reply by skliper** [reply]
There aren't APIs to do the segmentation for you, but get//set APIs are available for manipulating those fields.  See CFE_MSG_GetSegmentationFlag, CFE_MSG_SetSegmentationFlag, CFE_MSG_GetSequenceCount, CFE_MSG_SetSequenceCount at https://github.com/nasa/cFE/blob/c4ae5b262727f4f1381df596d00a6acc11c3a62c/modules/core_api/fsw/inc/cfe_msg.h#L229-L275.  If you are using an older version I think it provided macros, but I don't recall if it provided them for every field in the header.

**Reply by adamlin09** [reply]
Hi @skliper 

Thank you for your suggestion. I'm using CFE 6.7 now and I think CFE_MSG_ module hasn't been introduced yet. I will try to figure out and use the macros you mentioned if I need to interact with the message header.

**Comment by jwilmot** [comment, marked as answer]
On most (all?) flight processors, sending large messages is a significant performance hit, which is why it is not directly supported.  The common design pattern for  two or more applications to access the same large data set is shared memory with pointers and maybe a semaphore as needed. To transfer large data sets over a network, file transfer is used.  CFDP for Space system data links and FTP for IP networks.  You could also use files as the application to application sharing mechanism.

**Reply by astrogeco** [reply]
I second the file suggestion. @adamlin09 Is there a reason you have to avoid using files?

**Reply by adamlin09** [reply]
Hi @jwilmot and @astrogeco 

Thank you for your solution and suggestion! I just didn't find that cFS provide CF application for transferring and receiving files. After browsing the overview, it seems to be able to meet my requirement. I will try to use it to transfer and receive my images as files.

**Reply by astrogeco** [reply]
Glad to hear! Unfortunately the open-source version of CF isn't compatible with the current version of cFS. We will be releasing an update to CF soon though so stay tuned!
